src/diverse_seq/cluster.py

```diff
@@ -0,0 +1,601 @@
+    return sorted(-kmer_hash for kmer_hash in heap)
+
+
+def compute_mash_distance(
+    left_sketch: BottomSketch,
+    right_sketch: BottomSketch,
+    k: int,
+    sketch_size: int,
+) -> float:
+    """Compute the mash distance between two sketches.
+
+    Parameters
+    ----------
+    left_sketch : BottomSketch
+        A sketch for comparison.
+    right_sketch : BottomSketch
+        A sketch for comparison.
+    k : int
+        kmer size.
+    sketch_size : int
+        Size of the sketches.
+
+    Returns
+    -------
+    float
+        The mash distance between two sketches.
+    """
+    # Following the source code implementation
+    intersection_size = 0
+    union_size = 0
+
+    left_index = 0
+    right_index = 0
+    while (
+        union_size < sketch_size
+        and left_index < len(left_sketch)
+        and right_index < len(right_sketch)
+    ):
+        left, right = left_sketch[left_index], right_sketch[right_index]
+        if left < right:
+            left_index += 1
+        elif right < left:
+            right_index += 1
+        else:
+            left_index += 1
+            right_index += 1
+            intersection_size += 1
+        union_size += 1
+
+    if union_size < sketch_size:
+        if left_index < len(left_sketch):
+            union_size += len(left_sketch) - left_index
+        if right_index < len(right_sketch):
+            union_size += len(right_sketch) - right_index
+        union_size = min(union_size, sketch_size)
+
+    jaccard = intersection_size / union_size
+    if intersection_size == union_size:
+        return 0.0
+    if intersection_size == 0:
+        return 1.0
+    distance = -math.log(2 * jaccard / (1.0 + jaccard)) / k
+    if distance > 1:
+        distance = 1.0
+    return distance
+
+
+def iter_kmers(
+    seq: numpy.ndarray,
+    k: int,
+    num_states: int,
+) -> Generator[numpy.ndarray, None, None]:
+    """Iterator for kmers of a sequence.
+
+    Parameters
+    ----------
+    seq : numpy.ndarray
+        A sequence.
+    k : int
+        kmer size.
+    num_states : int
+        Number of states allowed for sequence type.
+
+    Yields
+    ------
+    numpy.ndarray
+        kmers of the sequence.
+    """
+    skip_until = 0
+    for i in range(k):
+        if seq[i] >= num_states:
+            skip_until = i + 1
+
+    for i in range(len(seq) - k + 1):
+        if seq[i + k - 1] >= num_states:
+            skip_until = i + k
+
+        if i < skip_until:
+            continue
+        yield seq[i : i + k]
+
+
+def hash_kmer(kmer: numpy.ndarray, *, mash_canonical: bool) -> int:
+    """Hash a kmer, optionally use the mash canonical representaiton.
+
+    Parameters
+    ----------
+    kmer : numpy.ndarray
+        The kmer to hash.
+    canonical : bool
+        Whether to use the mash canonical representation for a kmer.
+
+    Returns
+    -------
+    int
+        The has of a kmer.
+    """
+    tuple_kmer = tuple(map(int, kmer))
```

Comment:
FYI: you don't have to convert into a tuple unless you actually want to use it in a way that requires that
```python
import numpy

data = numpy.arange(1, 10, dtype=numpy.uint8)
hash(data.tobytes())
```

Reply:
This took me a bit to work out but converting to a tuple (or at least something that isn't a string/bytes string) is actually necessary when using multiprocessing to collect hashes. The result of `hash` is not guaranteed to be the same between different python processes for strings/byte strings ([they are salted before they are hashed to prevent some denial of service attacks](https://docs.python.org/3/reference/datamodel.html#object.__hash__)). Have kept the current method

Reply:
This took me a bit to work out but converting to a tuple (or at least something that isn't a string/bytes string) is actually necessary when using multiprocessing to collect hashes. The result of `hash` is not guaranteed to be the same between different python processes for strings/byte strings ([they are salted before they are hashed to prevent some denial of service attacks](https://docs.python.org/3/reference/datamodel.html#object.__hash__)). Have kept the current method